src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/DefaultShapeTableManager.cs

```diff
@@ -44,24 +47,30 @@ public class DefaultShapeTableManager : IShapeTableManager
             _logger = logger;
         }
 
-        public Task<ShapeTable> GetShapeTableAsync(string themeId)
+        public async Task<ShapeTable> GetShapeTableAsync(string themeId)
         {
             // This method is intentionally not awaited since most calls
             // are from cache.
-
             if (_shapeTableCache.TryGetValue(themeId ?? DefaultThemeIdKey, out var shapeTable))
             {
-                return Task.FromResult(shapeTable);
+                return shapeTable;
             }
 
-            lock (_shapeTableCache)
+            _semaphore ??= new SemaphoreSlim(1, 1);
+
+            await _semaphore.WaitAsync();
+            try
             {
                 if (_shapeTableCache.TryGetValue(themeId ?? DefaultThemeIdKey, out shapeTable))
                 {
-                    return Task.FromResult(shapeTable);
+                    return shapeTable;
                 }
 
-                return BuildShapeTableAsync(themeId);
+                return await BuildShapeTableAsync(themeId);
```

Comment:
Locking is only needed for this line, readers don't need a lock once the cache is warm.

Reply:
I see the happy path above, NVM.